### PR TITLE
ENH: Remove dependency upon python-apt library

### PR DESCRIPTION
**Pull request by rbuccigrossi**
The goal is to remove our dependency upon the python-apt library by using apt-cache calls.

- [x] Split _create_package into reasonable methods
- [x] Parse source_name and source_version (from Source: line if different than package)
- [x] Get the install_date for a package (from /var/lib/dpkg/info/_package_.list
- [x] Get the APTSource date from /var/lib/apt/lists/*Release file ("Date:" tag)
- [x] More unit testing

### CODE REVIEW

File: niceman/support/distributions/debian.py [hunk 1]
```diff
@@ -41,27 +38,36 @@ class DebianReleaseSpec(object):
...
-    release = deb822.Release(codecs.open(release_filename, 'r', 'utf-8'))
+    # RegExp to pull a single line "tag: value" pair from a deb822 file
+    re_deb822_single_line_tag = re.compile("""
+        ^(?P<tag>[a-zA-Z][^:]*):[\ ]+  # Tag - begins at start of line
+        (?P<val>\S.*)$           # Value - after colon to the end of the line
+    """, flags=re.VERBOSE + re.MULTILINE)
+
+    # Split before PGP signature if present
+    if "-----BEGIN PGP SIGNATURE-----" in content:
+        content = content.split("-----BEGIN PGP SIGNATURE-----")[0]
```
Comment (yarikoptic) [inline niceman/support/distributions/debian.py]: couldn't it be without 'if' statement and just the `content = content.split("-----BEGIN PGP SIGNATURE-----")[0]`? ;-)

File: niceman/support/distributions/debian.py [hunk 2]
```diff
@@ -71,30 +73,108 @@ def get_used_release_specs(package, installed_version=None):
...
+        ^(\s{5}|\s(?P<installed>\*\*\*)\s)
+        (?P<version>\S+)\s+
+        (?P<priority>\S+)\n.*
+        (?P<urls>(\n\s{8}.*)+)
+    """, flags=re.VERBOSE)
+    pkgs = {}
+    for entry in entries:
+        match = re_pkg.match(entry.strip())
+        if not match:
+            print("FAILED in ", entry)
```
[the remark this answers is not in the document]
Comment (rbuccigrossi) [reply niceman/support/distributions/debian.py]: ok

File: niceman/support/distributions/debian.py (hunk 1 above)
Comment (rbuccigrossi) [reply niceman/support/distributions/debian.py]: Look at that!  That's nice behavior for str.split()

**Comment by rbuccigrossi** [comment]
Assigned to Yarik for review and next steps.

**Comment by yarikoptic** [comment]
Eh, merging Matt's PR introduced some minor conflicts... I will resolve them now

**Comment by rbuccigrossi** [comment]
I'm merging to minimize future conflicts, but will resolve suggestions in a future pull request.

File: niceman/distributions/debian.py [hunk 3]
```diff
@@ -249,161 +229,208 @@ def _get_packagefields_for_files(self, files):
...
+                uri_suite):
+            try:
+                out = self._session.read(filename)
+                spec = get_spec_from_release_file(out)
+                try:
+                    date = str(pytz.utc.localize(
+                        datetime.utcfromtimestamp(
+                            mktime_tz(parsedate_tz(spec.date)))))
+                except TypeError as _:
+                    lgr.warning("Unexpected date format %s " % spec.date)
```
[the remark this answers is not in the document]
Comment (rbuccigrossi) [reply niceman/distributions/debian.py]: OK

File: niceman/distributions/debian.py [hunk 4]
```diff
@@ -249,161 +229,208 @@ def _get_packagefields_for_files(self, files):
...
+                break
+            except CommandError as _:
```
[the remark this answers is not in the document]
Comment (rbuccigrossi) [reply niceman/distributions/debian.py]: The issue is that we have two different potential file name endings ("Release" and "InRelease"), so we expect a bunch of failures in attempting to read the files.  I'll put better documentation in a comment.

File: niceman/tests/utils.py [hunk 5]
```diff
@@ -541,6 +542,20 @@ def newfunc(*args, **kwargs):
...
+def assert_is_subset_dict_recur(a, b):
+    """Asserts that 'a' is a subset of 'b' (recursively checks children dicts)
+    """
+    if isinstance(a, dict) and isinstance(b, dict):
+        for key in a:
+            if key not in b:
+                raise AssertionError("Key %s is missing" % key)
+            assert_is_subset_dict_recur(a[key], b[key])
+    else:
+        if not a == b:
```
[the remark this answers is not in the document]
Comment (rbuccigrossi) [reply niceman/tests/utils.py]: Yes. And, I love the fact that my assertion error used != while I used "not =="

File: niceman/resource/tests/test_shell.py [hunk 6]
```diff
@@ -72,6 +72,12 @@ def test_source_file_crash(script=None):
     with raises(Exception):  # TODO: unify?
         ses.source_script(script)
 
+@with_tempfile(content="content")
+def test_isdir(script=None):
+    ses = ShellSession()
+    assert not ses.isdir(script)
```
[the remark this answers is not in the document]
Comment (rbuccigrossi) [reply niceman/resource/tests/test_shell.py]: Done.

File: niceman/distributions/tests/test_debian.py [hunk 7]
```diff
@@ -8,20 +8,14 @@
...
 
 from pprint import pprint
 
 from niceman.distributions.debian import DebTracer
 
 import mock
-from niceman.tests.utils import skip_if
 
 
-@skip_if(not apt)
```
[the remark this answers is not in the document]
Comment (rbuccigrossi) [reply niceman/distributions/tests/test_debian.py]: Done.  I added a skip_if_no_apt_cache decorator in 7a6337b.

File: niceman/distributions/tests/test_debian.py [hunk 8]
```diff
@@ -102,6 +95,9 @@ def test_parse_dpkgquery_line():
     assert parse('fail2ban: /usr/bin/fail2ban-client') == \
            {'name': 'fail2ban', 'path': '/usr/bin/fail2ban-client'}
 
+    assert parse('fsl-5.0-eddy-nonfree, fsl-5.0-core: /usr/lib/fsl/5.0') == \
+           {'name': 'fsl-5.0-eddy-nonfree', 'path': '/usr/lib/fsl/5.0'}
+
     assert parse('diversion by dash from: /bin/sh') is None
```
[the remark this answers is not in the document]
Comment (rbuccigrossi) [reply niceman/distributions/tests/test_debian.py]: Done ( 496efe8 )

**Pull request merged**